File: features/logistic/shipping_recommendation/src/main/java/com/tokopedia/shipping_recommendation/shippingduration/view/ShippingDurationAdapter.java
package com.tokopedia.shipping_recommendation.shippingduration.view;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.tokopedia.shipping_recommendation.domain.shipping.ShippingDurationViewModel;

import java.util.List;

/**
 * Created by Irfan Khoirul on 08/08/18.
 */

public class ShippingDurationAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<ShippingDurationViewModel> shippingDurationViewModels;
    private ShippingDurationAdapterListener shippingDurationAdapterListener;
    private int cartPosition;
    // set true if has courier promo, whether own courier or other duration's courier
    private boolean hasCourierPromo;

    public void setShippingDurationViewModels(List<ShippingDurationViewModel> shippingDurationViewModels) {
        this.shippingDurationViewModels = shippingDurationViewModels;
    }

    public void setShippingDurationAdapterListener(ShippingDurationAdapterListener shippingDurationAdapterListener) {
        this.shippingDurationAdapterListener = shippingDurationAdapterListener;
    }

    public void setCartPosition(int cartPosition) {
        this.cartPosition = cartPosition;
    }

    public void setHasCourierPromo(boolean hasCourierPromo) {
        this.hasCourierPromo = hasCourierPromo;
    }

    @Override
    public int getItemViewType(int position) {
        return ShippingDurationViewHolder.ITEM_VIEW_SHIPMENT_DURATION;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        return new ShippingDurationViewHolder(view, this, cartPosition, hasCourierPromo);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ((ShippingDurationViewHolder) holder).bindData(shippingDurationViewModels.get(position),
                shippingDurationAdapterListener);
    }

    @Override
    public int getItemCount() {
        return shippingDurationViewModels.size();
    }

}

File: features/logistic/shipping_recommendation/src/main/java/com/tokopedia/shipping_recommendation/shippingduration/view/ShippingDurationAdapter.kt
package com.tokopedia.shipping_recommendation.shippingduration.view

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.tokopedia.shipping_recommendation.domain.shipping.LogisticPromoViewModel
import com.tokopedia.shipping_recommendation.domain.shipping.RatesViewModelType

import com.tokopedia.shipping_recommendation.domain.shipping.ShippingDurationViewModel

/**
 * Created by Irfan Khoirul on 08/08/18.
 */

class ShippingDurationAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var mData: MutableList<RatesViewModelType>
    private var shippingDurationAdapterListener: ShippingDurationAdapterListener? = null
    private var cartPosition: Int = 0
    // set true if has courier promo, whether own courier or other duration's courier
    private var hasCourierPromo: Boolean = false

    init {
        mData = mutableListOf()
    }

    fun setShippingDurationViewModels(shippingDurationViewModels: List<ShippingDurationViewModel>) {
        this.mData = shippingDurationViewModels.toMutableList()
        this.mData.add(0, LogisticPromoViewModel(1))
        notifyDataSetChanged()
    }

    fun setShippingDurationAdapterListener(shippingDurationAdapterListener: ShippingDurationAdapterListener) {
        this.shippingDurationAdapterListener = shippingDurationAdapterListener
    }

    fun setCartPosition(cartPosition: Int) {
        this.cartPosition = cartPosition
    }

    fun setHasCourierPromo(hasCourierPromo: Boolean) {
        this.hasCourierPromo = hasCourierPromo
    }

    override fun getItemViewType(position: Int): Int = when (mData.get(position)) {
        is LogisticPromoViewModel -> LogisticPromoViewHolder.LAYOUT
        else -> ShippingDurationViewHolder.ITEM_VIEW_SHIPMENT_DURATION
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(viewType, parent, false)
        return when (viewType) {
            LogisticPromoViewHolder.LAYOUT -> LogisticPromoViewHolder(view)
            else -> ShippingDurationViewHolder(view, this, cartPosition, hasCourierPromo)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is ShippingDurationViewHolder -> holder.bindData(mData.get(position) as ShippingDurationViewModel, shippingDurationAdapterListener)
            is LogisticPromoViewHolder -> holder.bindData(mData.get(position) as LogisticPromoViewModel)
        }
    }

    override fun getItemCount(): Int = mData.size

}
